### PR TITLE
Move version checking to handshake and deprecate Version RPC

### PR DESCRIPTION
**Pull request by xcchang**
## Overview

Description: In this PR we move the version checking to the RPC layer (the handshake handler), and added version info to RPC header such that we have the capability to detect when client and server are running different versions of corfu. The Version RPC has been removed since it was previously only used upon connection establish in CorfuRuntime. Unit test coverage for ClientHandshakeHandler and ServerHandshakeHandler is also added.

Why should this be merged: Done as part of the effort for rolling upgrade.

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: common/src/main/java/org/corfudb/common/CorfuSourceCodeVersion.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+    /**
+     * Getter for the long value which will be populated into our protobuf header. We only take the
+     * first {@link #CUTOFF} characters of {@link #CORFU_SOURCE_VERSION_STRING} as they
+     * are more than enough to uniquely identify a commit in our codebase.
+     *
+     * @return Long value converted from the first 12 characters of {@link #CORFU_SOURCE_VERSION_STRING}
+     */
+    public static long getCorfuSourceCodeVersion() {
+        return Long.parseLong(CORFU_SOURCE_VERSION_STRING.substring(0, CUTOFF), HEX_OPT);
```
Comment (xnull) [inline common/src/main/java/org/corfudb/common/CorfuSourceCodeVersion.java:33]: GitRepositoryState should be able to provide all needed information, please check this out https://github.com/CorfuDB/CorfuDB/blob/d42d17b360acfd2eccac4862b8f1fd1a79f75586/infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java#L433

Comment (xcchang) [reply common/src/main/java/org/corfudb/common/CorfuSourceCodeVersion.java:33]: Yeah we also explored that option, but we want a separate static definition because instead of updating this "version" every time the master branch get updated, we may just do it for every release or artifact deployment. This PR is still in draft mode and we are continue exploring automatic way to update the commit id here like using scripting, and it will be populated to other places like the handshake and layout.

File: runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java [hunk 2]
```diff
@@ -45,6 +42,18 @@ private CorfuProtocolMessage() {}
         IGNORE
     }
 
+    /**
+     * Returns a ProtocolVersionMsg with default capability_vector and static
+     * corfu_source_code_version.
```
Comment (chetangudisagar) [inline runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java:47]: As we are not setting any default value, change `default capability_vector` to `empty capability_vector list` maybe?

File: runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java [hunk 3]
```diff
@@ -45,6 +42,18 @@ private CorfuProtocolMessage() {}
...
+     *
+     * @return A ProtocolVersionMsg with default capability_vector.
```
Comment (chetangudisagar) [inline runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java:49]: same here.

File: common/src/main/java/org/corfudb/common/CorfuSourceCodeVersion.java (hunk 1 above)
Comment (hisundar) [reply common/src/main/java/org/corfudb/common/CorfuSourceCodeVersion.java:33]: GitRepositoryState isn't a library, it is a class written to extract certain generated files like "git.branch" etc and then populating it. These property files need to be correctly filled up for it to work.
What @xcchang is doing is essentially short-circuiting the process and making it directly available so it works in all the cases.
Automation to fill this can be added later.

Comment (xnull) [reply common/src/main/java/org/corfudb/common/CorfuSourceCodeVersion.java:33]: BTW, can we get corfu version from corfu app? Do we need anything else than the version? 
I mean we have "0.3.2.1" we know everything about the build, and release number?

I'm trying to understand the motivation here, it seems that ($major.$minor.$fix) semver works well in our case?

Comment (hisundar) [reply common/src/main/java/org/corfudb/common/CorfuSourceCodeVersion.java:33]: good question - it's just a safeguard to help folks with build to build breakages, if we retain major.minor then it becomes hard to test build-to-build upgrade failures, since these versions are not going to be modified in build-to-build setups. 
also having a git commit id in the version field protects us from the case where the version wasn't updated.

File: common/src/main/java/org/corfudb/common/CorfuSourceCodeVersion.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+/**
+ * Static class for representing the corfu source code version using git commit id.
+ */
+public final class CorfuSourceCodeVersion {
+
+    /**
+     * Original git commit id of the head of master branch, which should be updated upon every
+     * corfu artifact deployment.
+     */
+    private static final String CORFU_SOURCE_VERSION_STRING = "d42d17b360acfd2eccac4862b8f1fd1a79f75586";
```
Comment (chetangudisagar) [inline common/src/main/java/org/corfudb/common/CorfuSourceCodeVersion.java:13]: nit: can we rename it to `CORFU_SOURCE_CODE_VERSION_STRING` to keep it consistent with the class name?

File: infrastructure/src/main/java/org/corfudb/infrastructure/IServerRouter.java [hunk 5]
```diff
@@ -168,6 +169,12 @@ default boolean validateRequest(RequestMsg req, ChannelHandlerContext ctx) {
             return true;
         }
 
+        // TODO: In some cases this needs to be handled specially in the future
```
Comment (chetangudisagar) [inline infrastructure/src/main/java/org/corfudb/infrastructure/IServerRouter.java:172]: Could you please help me understand more as to in which cases we need to handle this especially?  Do you mean specific handling(throw error/print and return something) in the case of a version mismatch?

File: infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java [hunk 6]
```diff
@@ -106,14 +118,13 @@ public void channelRead(ChannelHandlerContext ctx, Object m) throws Exception {
         // Store clientID as a channel attribute.
         ctx.channel().attr(clientIdAttrKey).set(clientId);
         log.info("channelRead: Handshake validated by Server.");
-        log.debug("channelRead: Sending handshake response: Node Id: {} Corfu Version: {}",
-                this.nodeId, this.corfuVersion);
+        log.debug("channelRead: Sending handshake response: Node Id: {}", this.nodeId);
```
Comment (chetangudisagar) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java:121]: Can we print a debug message of the server version like before? Might help in bug triaging as this is the only place where we are doing main version checks.

```
log.debug("channelRead: Sending handshake response: Node Id: {}, ServerVersion: {}, ClientVersion: {}", this.nodeId, serverVersion.toString(), clientVersion.toString());
````

File: infrastructure/src/test/java/org/corfudb/infrastructure/BaseServerTest.java [hunk 7]
```diff
@@ -182,37 +177,6 @@ public void testSealWrongEpoch() {
...
-        assertTrue(compareBaseHeaderFields(request.getHeader(), response.getHeader()));
-        assertTrue(response.getPayload().hasVersionResponse());
-
-        // Assert that the contents of the versionInfo are what we expect
-        VersionInfo versionInfo = getVersionInfo(response.getPayload().getVersionResponse());
-        assertEquals("localhost", versionInfo.getNodeId());
-        assertEquals(1, versionInfo.getOptionsMap().size());
-        assertEquals("localhost", versionInfo.getOptionsMap().get("--address"));
-    }
-
```
Comment (chetangudisagar) [inline infrastructure/src/test/java/org/corfudb/infrastructure/BaseServerTest.java:215]: If required, can we write some test cases to check that the warning message was printed when there is a version mismatch and was not printed when there is no mismatch? Later this can be extended to checking the thrown exceptions or printing some other error/warn messages.

File: runtime/src/main/java/org/corfudb/util/GitRepositoryState.java [hunk 8]
```diff
@@ -76,4 +76,19 @@ public static GitRepositoryState getRepositoryState() {
...
+     * @return Long value converted from the first 12 characters of parent git commit id.
+     */
+    public static long getCorfuSourceCodeVersion() {
+        if (_gitRepositoryState == null) {
+            getRepositoryState();
+        }
+        final int CUTOFF = 12;
+        final int HEX_OPT = 16;
+        return Long.parseLong(_gitRepositoryState.commitId.substring(0, CUTOFF), HEX_OPT);
+    }
```
Comment (chetangudisagar) [inline runtime/src/main/java/org/corfudb/util/GitRepositoryState.java:93]: Maybe we can write some unit tests for this method too.

File: infrastructure/src/main/java/org/corfudb/infrastructure/IServerRouter.java (hunk 5 above)
Comment (xcchang) [reply infrastructure/src/main/java/org/corfudb/infrastructure/IServerRouter.java:172]: I thought currently this is mainly used for logging and as a placeholder. From our discussion before, I remember we want to forward requestMsg with different versions to different versions of handlers/servers (for example, `Sequencer ServerV2` or `handleTokenRequestV2`, this part of design is still pending to be determined though)

**Review by zfrenette — COMMENTED** [review]
First batch of comments -- Can we add tests in this PR for these versioning changes?

File: infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java (hunk 6 above)
Comment (xcchang) [reply infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java]: Added, thanks!

File: test/src/test/java/org/corfudb/runtime/clients/BaseHandlerTest.java [hunk 9]
```diff
@@ -32,9 +32,4 @@
                 .add(baseHandler)
                 .build();
     }
-
-    @Test
-    public void canGetVersionInfo() {
-        CFUtils.getUninterruptibly(client.getVersionInfo());
-    }
```
[the remark this answers is not in the document]
Comment (xcchang) [reply test/src/test/java/org/corfudb/runtime/clients/BaseHandlerTest.java:39]: Removed, thanks!

File: infrastructure/src/main/java/org/corfudb/infrastructure/IServerRouter.java [hunk 10]
```diff
@@ -168,6 +169,12 @@ default boolean validateRequest(RequestMsg req, ChannelHandlerContext ctx) {
...
+        if (requestHeader.getVersion().getCorfuSourceCodeVersion() !=
+                GitRepositoryState.getCorfuSourceCodeVersion()) {
+            log.warn("validateRequest: Version mismatch.");
```
[the remark this answers is not in the document]
Comment (xcchang) [reply infrastructure/src/main/java/org/corfudb/infrastructure/IServerRouter.java]: Added, thanks!

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServerNode.java [hunk 11]
```diff
@@ -324,8 +322,6 @@ protected void initChannel(@Nonnull Channel ch) throws Exception {
                 ch.pipeline().addLast(new NettyCorfuMessageDecoder());
                 ch.pipeline().addLast(new NettyCorfuMessageEncoder());
                 ch.pipeline().addLast(new ServerHandshakeHandler(context.getNodeId(),
-                        Version.getVersionString() + "("
```
[the remark this answers is not in the document]
Comment (xcchang) [reply infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServerNode.java:327]: Done, thanks!

File: infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java [hunk 12]
```diff
@@ -67,16 +65,30 @@ public void channelRead(ChannelHandlerContext ctx, Object m) throws Exception {
...
+            log.warn("channelRead: Dropping the message as the received msg is not an instance" +
+                    " of RequestMsg. Message - {}", m);
+            return;
+        }
+
+        RequestMsg requestMsg = (RequestMsg) m;
+        long clientVersion = requestMsg.getHeader().getVersion().getCorfuSourceCodeVersion();
+        long serverVersion = GitRepositoryState.getCorfuSourceCodeVersion();
+        if (clientVersion != serverVersion) {
+            log.warn("channelRead: Version mismatch. Corfu client version: {}, Corfu server version: {}",
```
[the remark this answers is not in the document]
Comment (xcchang) [reply infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java:78]: Done, thanks!

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java [hunk 13]
```diff
@@ -129,7 +130,11 @@ public void channelRead(ChannelHandlerContext ctx, Object m) throws Exception {
             return;
         }
 
-        log.info("channelRead: Handshake succeeded. Server Corfu Version: [{}]", corfuVersion);
+        log.info("channelRead: Handshake succeeded. Corfu Server Version: [{}]", corfuServerVersion);
+        if (corfuServerVersion != GitRepositoryState.getCorfuSourceCodeVersion()) {
+            log.warn("channelRead: Version mismatch. Corfu client version: {}, Corfu server version: {}",
+                    GitRepositoryState.getCorfuSourceCodeVersion(), corfuServerVersion);
```
[the remark this answers is not in the document]
Comment (xcchang) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java]: Done, thanks!

File: infrastructure/src/test/java/org/corfudb/infrastructure/BaseServerTest.java (hunk 7 above)
Comment (xcchang) [reply infrastructure/src/test/java/org/corfudb/infrastructure/BaseServerTest.java:215]: Unit tests added, thanks!

File: common/src/main/java/org/corfudb/common/CorfuSourceCodeVersion.java (hunk 1 above)
Comment (xcchang) [reply common/src/main/java/org/corfudb/common/CorfuSourceCodeVersion.java]: Note: We had off-line discussions and found the `git.properties` file is actually carried into `runtime.jar` and `infrastructure.jar`, so it could be good to use the `GitRepositoryState`.

File: infrastructure/src/main/java/org/corfudb/infrastructure/IServerRouter.java (hunk 5 above)
Comment (xcchang) [reply infrastructure/src/main/java/org/corfudb/infrastructure/IServerRouter.java]: Note: We finally decide to remove this logging here to avoid flooding the logs.

File: runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java [hunk 14]
```diff
@@ -45,6 +42,18 @@ private CorfuProtocolMessage() {}
...
 
+    /**
+     * Returns a ProtocolVersionMsg with empty capability_vector and static
+     * corfu_source_code_version.
+     *
+     * @return A ProtocolVersionMsg with empty capability_vector.
+     */
+    public static ProtocolVersionMsg getDefaultProtocolVersionMsg() {
+        return ProtocolVersionMsg.newBuilder()
+                .setCorfuSourceCodeVersion(GitRepositoryState.getCorfuSourceCodeVersion())
```
Comment (zfrenette) [inline runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java:53]: Would it be better to initialize the compatibility vector as well? It seems like right now we are using an empty list to denote the default capabilities instead of a vector of all zeros.

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 15]
```diff
@@ -1294,8 +1244,6 @@ public synchronized CorfuRuntime connect() {
             }
         }
 
-        checkVersion();
```
Comment (zfrenette) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java:1297]: How will the runtime become aware of the version of the nodes in the cluster? Will it need to?

File: runtime/pom.xml [hunk 16]
```diff
@@ -173,6 +173,12 @@
             <version>6.0.53</version>
             <scope>provided</scope>
         </dependency>
+        <dependency>
+            <groupId>ch.qos.logback</groupId>
+            <artifactId>logback-classic</artifactId>
+            <version>${logback.classic.version}</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (zfrenette) [inline runtime/pom.xml:181]: Why is this needed? I believe it was removed because of conflicts in NSX:  https://github.com/CorfuDB/CorfuDB/pull/2841/.

Comment (xcchang) [reply runtime/pom.xml:181]: It is for inspecting the logs in unit test of `ClientHandshakeHandler`. I set the scope to be test hoping to avoid the issue, but I will double check.

File: runtime/src/test/java/org/corfudb/runtime/ClientHandshakeHandlerTest.java [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+     * with default values.
+     *
+     * @param ignoreClusterId   indicates if the message is clusterId aware
+     * @param ignoreEpoch       indicates if the message is epoch aware
+     * @return                  the corresponding HeaderMsg
+     */
+    private HeaderMsg getBasicHeader(ClusterIdCheck ignoreClusterId, EpochCheck ignoreEpoch) {
+        return getHeaderMsg(requestCounter.incrementAndGet(), PriorityLevel.NORMAL, 0L,
+                getUuidMsg(DEFAULT_UUID), getUuidMsg(DEFAULT_UUID), ignoreClusterId, ignoreEpoch);
+    }
```
Comment (zfrenette) [inline runtime/src/test/java/org/corfudb/runtime/ClientHandshakeHandlerTest.java:64]: We should consider factoring/restructuring common RPC unit test code in a subsequent PR.

File: runtime/src/test/java/org/corfudb/runtime/ClientHandshakeHandlerTest.java [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+                SERVER_NODEID, DEFAULT_HANDSHAKE_TIMEOUT);
+        embeddedChannel = new EmbeddedChannel(clientHandshakeHandler);
+
+        // Setting up the logger such that we could inspect the logged messages in ClientHandshakeHandler.
+        Logger logger = (Logger) LoggerFactory.getLogger(ClientHandshakeHandler.class);
+        logger.setLevel(Level.DEBUG);
+
+        listAppender = new ListAppender<>();
+        listAppender.start();
+        logger.addAppender(listAppender);
```
Comment (zfrenette) [inline runtime/src/test/java/org/corfudb/runtime/ClientHandshakeHandlerTest.java:78]: I think relying on logging for unit tests makes them fragile. Handshake success/failure can be tested by detecting the presence or absence of another message sent down the pipeline. In the case of a version mismatch, maybe there isn't enough substance for a unit test, since it has no impact on the behaviour yet.

File: infrastructure/src/test/java/org/corfudb/infrastructure/ServerHandshakeHandlerTest.java [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+     */
+    private HeaderMsg getBasicHeader(ClusterIdCheck ignoreClusterId, EpochCheck ignoreEpoch) {
+        return getHeaderMsg(requestCounter.incrementAndGet(), PriorityLevel.NORMAL, 0L,
+                getUuidMsg(DEFAULT_UUID), getUuidMsg(DEFAULT_UUID), ignoreClusterId, ignoreEpoch);
+    }
+
+    @Before
+    public void setup() {
+        ServerHandshakeHandler serverHandshakeHandler = new ServerHandshakeHandler(SERVER_NODEID,
+                GitRepositoryState.getCorfuSourceCodeVersion(), DEFAULT_HANDSHAKE_TIMEOUT);
```
Comment (zfrenette) [inline infrastructure/src/test/java/org/corfudb/infrastructure/ServerHandshakeHandlerTest.java:67]: Since this is a unit test, maybe the server version can be mocked as well so it doesn't have to rely on GitRepositoryState.

File: runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java [hunk 20]
```diff
@@ -145,6 +165,14 @@ public static ResponseMsg getResponseMsg(HeaderMsg header, ResponsePayloadMsg re
      * @return        a ResponseMsg with the provided header and error
      */
     public static ResponseMsg getResponseMsg(HeaderMsg header, ServerErrorMsg error) {
+        if (header.getVersion().getCorfuSourceCodeVersion() !=
+                GitRepositoryState.getCorfuSourceCodeVersion()) {
```
Comment (zfrenette) [inline runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java:169]: Is this the best place for such a check? The assumption here was that the response header was already built.

File: runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java [hunk 21]
```diff
@@ -131,6 +143,14 @@ public static RequestMsg getRequestMsg(HeaderMsg header, RequestPayloadMsg reque
      * @return           a ResponseMsg with the provided header and payload
      */
     public static ResponseMsg getResponseMsg(HeaderMsg header, ResponsePayloadMsg response) {
+        if (header.getVersion().getCorfuSourceCodeVersion() !=
+                GitRepositoryState.getCorfuSourceCodeVersion()) {
```
Comment (zfrenette) [inline runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java:147]: Maybe this can be a private constant in the API class.

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java [hunk 22]
```diff
@@ -95,6 +96,13 @@ public void channelRead(ChannelHandlerContext ctx, Object m) throws Exception {
         }
 
         ResponseMsg response = ((ResponseMsg) m);
+        long corfuServerVersion = response.getHeader().getVersion().getCorfuSourceCodeVersion();
+        long corfuClientVersion = GitRepositoryState.getCorfuSourceCodeVersion();
+
+        if (corfuServerVersion != GitRepositoryState.getCorfuSourceCodeVersion()) {
```
Comment (zfrenette) [inline runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java:102]: `corfuServerVersion != corfuClientVersion`?

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java (hunk 15 above)
Comment (xcchang) [reply runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java:1297]: Currently the server version is contained in the response header, and for now only the ClientHandshakeHandler will be aware of it.

File: runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java (hunk 14 above)
Comment (xcchang) [reply runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java:53]: From our discussion this part would be handled in a separate PR for capability vector implementation.

File: infrastructure/src/test/java/org/corfudb/infrastructure/ServerHandshakeHandlerTest.java (hunk 19 above)
Comment (xcchang) [reply infrastructure/src/test/java/org/corfudb/infrastructure/ServerHandshakeHandlerTest.java]: Changed, thanks!

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java (hunk 22 above)
Comment (xcchang) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java]: Done, thanks!

File: runtime/pom.xml (hunk 16 above)
Comment (xcchang) [reply runtime/pom.xml]: As we no longer inspect log message in test cases, it is now removed, thanks!

File: runtime/src/test/java/org/corfudb/runtime/ClientHandshakeHandlerTest.java (hunk 18 above)
Comment (xcchang) [reply runtime/src/test/java/org/corfudb/runtime/ClientHandshakeHandlerTest.java]: Changed, thanks!

File: runtime/src/test/java/org/corfudb/runtime/ClientHandshakeHandlerTest.java (hunk 17 above)
Comment (xcchang) [reply runtime/src/test/java/org/corfudb/runtime/ClientHandshakeHandlerTest.java:68]: Will do this in a following cleanup/refactor PR.

File: runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java (hunk 20 above)
Comment (xcchang) [reply runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java:171]: Will do this in a following cleanup/refactor PR.

File: runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java (hunk 21 above)
Comment (xcchang) [reply runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java:149]: Will do this in a following cleanup/refactor PR.

File: runtime/proto/service/corfu_message.proto [hunk 23]
```diff
@@ -22,7 +22,8 @@ enum PriorityLevel {
 }
 
 message ProtocolVersionMsg {
-  fixed64 capability_vector = 1;
+  repeated fixed64 capability_vector = 1;
+  int64 corfu_source_code_version = 2;
```
Comment (hisundar) [inline runtime/proto/service/corfu_message.proto:26]: maybe a comment explaining how this is populated?

File: test/src/test/resources/logback-test.xml [hunk 24]
```diff
@@ -61,7 +61,7 @@
     <!-- Control logging levels for individual components here. -->
     <logger name="org.corfudb.runtime.object" level="INFO"/>
     <logger name="org.corfudb.runtime.clients" level="INFO"/>
-    <logger name="org.corfudb.infrastructure" level="INFO"/>
+    <logger name="org.corfudb.infrastructure" level="WARN"/>
```
Comment (hisundar) [inline test/src/test/resources/logback-test.xml:64]: why did we change the test logging to WARN?

File: test/src/test/resources/logback-test.xml [hunk 25]
```diff
@@ -78,7 +78,7 @@
 
     <root level="INFO">
         <!--<appender-ref ref="FILE" />-->
-        <!--<appender-ref ref="STDOUT" />-->
+        <appender-ref ref="STDOUT" />
```
Comment (hisundar) [inline test/src/test/resources/logback-test.xml:81]: Can you please remove these changes to the logback since it would create a lot of noise in the GitHub action verifications

**Review by hisundar — COMMENTED** [review]
Looks good @xcchang just a few minor comments.

File: test/src/test/resources/logback-test.xml (hunk 24 above)
Comment (xcchang) [reply test/src/test/resources/logback-test.xml:64]: Sorry for the confusion, those from the second commit are just temporary for debugging the flaky test.

**Review by hisundar — APPROVED** [review]
LGTM, please remove the test changes once the tests run fine

File: runtime/src/main/java/org/corfudb/util/GitRepositoryState.java (hunk 8 above)
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/util/GitRepositoryState.java:93]: This has a race condition. Writing to `_gitRepositoryState` and accessing it can happen from multiple threads but it's not synchronized. 

`_gitRepositoryState` can be made a singleton.

File: test/src/test/java/org/corfudb/CorfuTestParameters.java [hunk 26]
```diff
@@ -87,8 +87,8 @@ public CorfuTestParameters(){
 
         // Timeouts
         TIMEOUT_VERY_SHORT = Duration.of(500, MILLIS);
-        TIMEOUT_SHORT = Duration.of(1, SECONDS);
-        TIMEOUT_NORMAL = Duration.of(10, SECONDS);
+        TIMEOUT_SHORT = Duration.of(5, SECONDS);
+        TIMEOUT_NORMAL = Duration.of(60, SECONDS);
         TIMEOUT_LONG = Duration.of(2, MINUTES);
```
Comment (Maithem) [inline test/src/test/java/org/corfudb/CorfuTestParameters.java:92]: This should be part of a separate PR.

File: runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java (hunk 20 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java:171]: +1 @zfrenette  comment. I think it should be addressed in this PR.

File: test/src/test/java/org/corfudb/CorfuTestParameters.java (hunk 26 above)
Comment (xcchang) [reply test/src/test/java/org/corfudb/CorfuTestParameters.java:92]: Yeah this will be removed, it's temporary for debugging a flaky test, sorry for the confusion

File: runtime/proto/service/corfu_message.proto (hunk 23 above)
Comment (Maithem) [reply runtime/proto/service/corfu_message.proto:26]: Please address this.

**Review by Maithem — APPROVED** [review]

File: runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java (hunk 20 above)
Comment (xcchang) [reply runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java]: Done, thanks!

File: runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java (hunk 21 above)
Comment (xcchang) [reply runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolMessage.java]: Done, thanks!

File: runtime/src/main/java/org/corfudb/util/GitRepositoryState.java (hunk 8 above)
Comment (xcchang) [reply runtime/src/main/java/org/corfudb/util/GitRepositoryState.java]: Revised, thanks a lot for pointing out this!
